### ISSUE TITLE
Instance-level retrieval (Flickr30K)

### ISSUE THREAD

**Issue by MikelWL** [penghu-cs/UCCH#19]
Thank you for your very interesting work!

I am currently working on seeing how category-level Cross-Modal Hashing models compare to instance-level fine-grained retrieval models, in an effort to quantify how much of a speed/storage gain is possible via CMH contrasted with how much of a sacrifice there is on recall/precision. The results for instance-retrieval on Flickr30K from Table 4 of the paper show interesting potential, however, I am not able to reproduce them.

The approach I am currently attempting is rather straightforward: Taking the already existing training and evaluation functionality as well as following the same dataset pre-processing, and finally implementing recall@k to the same hash codes on which the existing mAP evaluation is calculated. I see that the paper mentions that you implant your model onto the VSE++ framework for fair comparison. I am wondering what is meant by this in more detail, and what steps within this process may have led to the promising results showcased in the paper compared to my straightforward attempt.

Any other additional insights on the topic of applying your model to the instance-retrieval task would be greatly appreciated.

Many thanks in advance.

**Comment by penghu-cs** [comment]
Thank you for reaching out and for your interest in our work. To address your query regarding the reproduction of results from Table 4 for instance-retrieval on Flickr30K, let me provide some additional details.

In our approach, we indeed utilize the VSE++ framework as a baseline for fair comparison. Specifically, we integrate our model into this framework by following the code available at [VSE++ GitHub repository](https://github.com/fartashf/vsepp.git). During the training stage, we substitute the original ContrastiveLoss with our UCCH loss. This modification is crucial as it aligns the training objectives with the hashing-based retrieval task. For the evaluation in the test stage, we perform cross-modal retrieval using the binary codes generated by our model. This means that instead of using the raw feature representations, we use the hash codes to compute the recall@k metric, which is a standard practice for evaluating retrieval performance in binary code-based systems. If you're following a similar approach but are not observing the expected results, it might be helpful to double-check the implementation details, especially the loss function and the binary code generation process.

We hope this clarifies the steps we took and assists you in your endeavors. If you have further questions or require additional insights on applying our model to the instance-retrieval task, please feel free to ask.

Best,
Peng

**Comment by MikelWL** [comment]
Thank you very much for your prompt reply.

I would like to attempt to follow your approach of implanting UCCH onto the VSE++ framework, however, due to some time constraints with my current work I do not think I will be able to achieve this after a brief study of estimating how long this would take me.

The most realistic solution given time constraints would be for me to correctly implement instance-level recall evaluation on top of the UCCH code. However, assuming this is done correctly, are there any major obvious reasons why this may not work compared to the implantation of UCCH onto VSE++? I.e., does performing this implantation inherit crucial instance-level retrieval capabilities from VSE++ which UCCH wouldn't have otherwise? Or should the current standalone implementation of UCCH in theory be able to produce somewhat similar results to those obtained when implanted onto VSE++?

I am very grateful for any insights you may be able to provide.

Best regards,
Mikel

**Comment by penghu-cs** [comment]
Dear Mikel,

Thank you for reaching out with your query.

Implementing instance-level recall evaluation on top of the UCCH code is indeed a feasible approach. While the standalone UCCH implementation is capable of producing comparable results, integrating its loss into the VSE++ framework can reduce the time required for reproduction, given the time constraints you mentioned. 

Please find the computational codes for image-to-text (i2t) and text-to-image (t2i) evaluations attached for your reference. These should aid in your implementation.

I hope this information is helpful to you. Should you have any further questions, please do not hesitate to ask.

Warm regards,
Peng

```python
def i2t(images, captions, npts=None, measure='hamming', return_ranks=False):
    """
    Images->Text (Image Annotation)
    Images: (5N, K) matrix of images
    Captions: (5N, K) matrix of captions
    """
    if npts is None:
        npts = int(images.shape[0] / 5)
    index_list = []

    ranks = numpy.zeros(npts)
    top1 = numpy.zeros(npts)
    for index in range(npts):

        # Get query image
        im = images[5 * index].reshape(1, images.shape[1])

        # Compute scores
        if measure == 'order':
            bs = 100
            if index % bs == 0:
                mx = min(images.shape[0], 5 * (index + bs))
                im2 = images[5 * index:mx:5]
                d2 = order_sim(torch.Tensor(im2).cuda(),
                               torch.Tensor(captions).cuda())
                d2 = d2.cpu().numpy()
            d = d2[index % bs]
        elif measure == 'cosine':
            d = numpy.dot(im, captions.T).flatten()
        else:
            # d = numpy.dot(im, captions.T).flatten()
            d = -scipy.spatial.distance.cdist(im, captions, measure).reshape([-1])
        inds = numpy.argsort(d)[::-1]
        index_list.append(inds[0])

        # Score
        rank = 1e20
        for i in range(5 * index, 5 * index + 5, 1):
            tmp = numpy.where(inds == i)[0][0]
            if tmp < rank:
                rank = tmp
        ranks[index] = rank
        top1[index] = inds[0]

    # Compute metrics
    r1 = 100.0 * len(numpy.where(ranks < 1)[0]) / len(ranks)
    r5 = 100.0 * len(numpy.where(ranks < 5)[0]) / len(ranks)
    r10 = 100.0 * len(numpy.where(ranks < 10)[0]) / len(ranks)
    medr = numpy.floor(numpy.median(ranks)) + 1
    meanr = ranks.mean() + 1
    if return_ranks:
        return (r1, r5, r10, medr, meanr), (ranks, top1)
    else:
        return (r1, r5, r10, medr, meanr)


def t2i(images, captions, npts=None, measure='hamming', return_ranks=False):
    """
    Text->Images (Image Search)
    Images: (5N, K) matrix of images
    Captions: (5N, K) matrix of captions
    """
    if npts is None:
        npts = int(images.shape[0] / 5)
    ims = numpy.array([images[i] for i in range(0, len(images), 5)])

    ranks = numpy.zeros(5 * npts)
    top1 = numpy.zeros(5 * npts)
    for index in range(npts):

        # Get query captions
        queries = captions[5 * index:5 * index + 5]

        # Compute scores
        if measure == 'order':
            bs = 100
            if 5 * index % bs == 0:
                mx = min(captions.shape[0], 5 * index + bs)
                q2 = captions[5 * index:mx]
                d2 = order_sim(torch.Tensor(ims).cuda(),
                               torch.Tensor(q2).cuda())
                d2 = d2.cpu().numpy()

            d = d2[:, (5 * index) % bs:(5 * index) % bs + 5].T
        elif measure == 'cosine':
            d = numpy.dot(queries, ims.T)
        else:
            # d = numpy.dot(queries, ims.T)
            d = -scipy.spatial.distance.cdist(queries, ims, measure)
        inds = numpy.argsort(d)[::-1]
        inds = numpy.zeros(d.shape)
        for i in range(len(inds)):
            inds[i] = numpy.argsort(d[i])[::-1]
            ranks[5 * index + i] = numpy.where(inds[i] == index)[0][0]
            top1[5 * index + i] = inds[i][0]

    # Compute metrics
    r1 = 100.0 * len(numpy.where(ranks < 1)[0]) / len(ranks)
    r5 = 100.0 * len(numpy.where(ranks < 5)[0]) / len(ranks)
    r10 = 100.0 * len(numpy.where(ranks < 10)[0]) / len(ranks)
    medr = numpy.floor(numpy.median(ranks)) + 1
    meanr = ranks.mean() + 1
    if return_ranks:
        return (r1, r5, r10, medr, meanr), (ranks, top1)
    else:
        return (r1, r5, r10, medr, meanr)
```
